bcda/models/postgres/alr.go

```diff
@@ -0,0 +1,252 @@
+	if err := enc.Encode(mp); err != nil {
+		return nil, err
+	}
+	return bytea.Bytes(), nil
+}
+
+func GobDecoder(byteMap []byte) (map[string]string, error) {
+	var mp map[string]string
+	r := bytes.NewReader(byteMap)
+	dec := gob.NewDecoder(r)
+	if err := dec.Decode(&mp); err != nil {
+		return nil, err
+	}
+	return mp, nil
+}
+
+/*******************************************************************************
+	METHODS
+		1. Next, Err, Values
+			- These methods are need to implement the CopyFrom from pgx.
+		2. AddAlr
+			- Adds new rows to the ALR Tables. A single call of this function
+			can ingest ALR data per ACO.
+		3. GetAlr
+			- Retreive data from database.
+*******************************************************************************/
+
+func (a *alrCopyFromSource) Next() bool {
+	// If we went through all of the rows, this condition should be false
+	return a.alrCount < a.alrRowLen
+}
+
+func (a *alrCopyFromSource) Err() error {
+	// Err() returns nil if "Done" has not closed and context is still valid
+	return a.ctx.Err()
+}
+
+func (a *alrCopyFromSource) Values() ([]interface{}, error) {
+	// Get the row
+	alr := a.alrs[a.alrCount]
+	// Move counter up so we look at the following row next
+	a.alrCount++
+
+	// Convert the map[string]string to slices of bytes
+	bytea, err := GobEncoder(alr.KeyValue)
+	if err != nil {
+		return nil, err
+	}
+
+	row := []interface{}{
+		// Ordering is based off how they are in the database. See migration 10
+		a.metaKey,   // bigint
+		alr.BeneMBI, // varchar
+		alr.BeneHIC,
+		alr.BeneFirstName,
+		alr.BeneLastName,
+		alr.BeneSex,
+		alr.BeneDOB,
+		alr.BeneDOD,
+		bytea, // bytea
+	}
+	return row, nil
+}
+
+func (r *AlrRepository) AddAlr(ctx context.Context, aco string, timestamp time.Time, alrs []models.Alr) error {
+	// Do this in a single transaction using BeginEX from the pgx package
+	tx, err := r.BeginEx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// Update the alr_meta table and get the foreign key
+	updateAlrMeta := sqlFlavor.NewInsertBuilder().InsertInto("alr_meta")
+	updateAlrMeta.Cols("aco", "timestp").Values(aco, timestamp)
+	query, args := updateAlrMeta.Build()
+	query = fmt.Sprintf("%s RETURNING id", query)
+
+	var metaKey int
+	if err := tx.QueryRowEx(ctx, query, nil, args...).Scan(&metaKey); err != nil {
+		return err
+	}
+
+	// Update the alr table
+	cfs := &alrCopyFromSource{
+		ctx:       ctx,
+		metaKey:   metaKey,
+		timestamp: timestamp,
+		alrCount:  0,
+		alrRowLen: len(alrs),
+		alrs:      alrs,
+	}
+
+	fields := []string{"metakey", "mbi", "hic", "firstname", "lastname", "sex", "dob", "dod", "keyvalue"}
+	_, err = tx.CopyFrom(pgx.Identifier([]string{"alr"}), fields, cfs)
+	// If the copyfrom fails, attempt to rollback changes
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			logrus.Warnf("Failed to rollback transaction %s", err.Error())
+		}
+		return err
+	}
+
+	// End the transaction
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+	return nil
```

Comment:
Can we put this rollback/commit logic in a defer statement? After we've successfully started the transaction.